[CMake] Make GNU-style response files for long argument lists

jrose-apple opened the pull request:
...i.e. an actual shell-like argument list, rather than a semicolon-separated list (CMake's internal stringification mechanism for lists). Apart from being a little easier to read, this also works directly with the response file support in swiftc itself now (not depending on utils/line-directive).

(It's still not *quite* enough to expand on a command-line, though, since that will escape the quotes. The 'sed' command can get around that: `$(sed "s/'//g" foo.txt)`.)

jrose-apple commented:
@swift-ci Please smoke test

compnerd commented on cmake/modules/SwiftSource.cmake:
```diff
@@ -455,13 +455,14 @@ function(_compile_swift_files
   set(file_path "${CMAKE_CURRENT_BINARY_DIR}/${file_name}.txt")
-  file(WRITE "${file_path}" "${source_files}")
+  string(REPLACE ";" "'\n'" source_files_quoted "${source_files}")
+  file(WRITE "${file_path}" "'${source_files_quoted}'")
```
Im worried about how this expansion works on Windows.  The shell escaping there is weird and the quoting here I believe will go wrong.  Note that `shlex` behaves differently on Windows and unices, so the behaviour of the quoting is relevant to that too

jrose-apple replied:
One of `shlex`'s arguments is whether it behaves POSIXly, which I passed `True` to. Still, I'd love to test this on a Windows machine; I just don't have one available today.

Rostepher commented on utils/line-directive:
```diff
@@ -178,7 +179,10 @@ def map_line_from_source_file(source_filename, source_line_num,
+        # "Make an iterator out of shlex.shlex.get_token, then consume items
+        # until it returns None." (Then eagerly convert the result to a list so
+        # that we can close the file.)
+        return list(iter(shlex.shlex(files, file_path, True).get_token, None))
```
I'd prefer for the `True` value we use a named argument, if that's `shell` it makes it much easier to understand what's going on here.

jrose-apple replied:
I'd love to as well; unfortunately [shlex doesn't document a name](https://docs.python.org/2/library/shlex.html#shlex.shlex).

Rostepher replied:
I believe the name is `posix`, the function signature is `shlex.shlex(instream=None, infile=None, posix=False)` [Lib/shlex.py](https://github.com/python/cpython/blob/491740f116755e220135e596ec802ea3a0f65596/Lib/shlex.py#L23)

jrose-apple committed:
[CMake] Make GNU-style response files for long argument lists

...i.e. an actual shell-like argument list, rather than a
semicolon-separated list (CMake's internal stringification mechanism
for lists). Apart from being a little easier to read, this also works
directly with the response file support in swiftc itself now (not
depending on utils/line-directive).

(It's still not /quite/ enough to expand on a command-line, though,
since that will escape the quotes. The 'sed' command can get around
that: $(sed "s/'//g" foo.txt).)

jrose-apple commented:
@swift-ci Please clean smoke test

jrose-apple commented:
@Rostepher @compnerd ping

Rostepher commented:
This looks good to me, I have no further comments nor suggestions. 👍

compnerd commented:
@jrose-apple - sorry, forgot about this.  Lets go ahead and merge this.  If there are problems with Windows, we can fix them then.

jrose-apple commented:
Let's make sure this still integrates cleanly.

@swift-ci Please smoke test
